File: solutions/typescript/variable-length-quantity/1/variable-length-quantity.ts
export function encode(numbers: number[]): number[] {
  if (!Array.isArray(numbers)) {
    throw new Error('Input must be array');
  }
  
  const result: number[] = [];
  
  for (let num of numbers) {
    if (num < 0 || num > 0xFFFFFFFF) {
      throw new Error('Number out of range (must be 32-bit unsigned integer)');
    }
    
    if (num === 0) {
      result.push(0);
      continue;
    }
    
    const chunks: number[] = [];
    let n = num;
    
    while (n > 0) {
      chunks.unshift(n & 0x7F);
      n >>>= 7;
    }
    
    for (let i = 0; i < chunks.length; i++) {
      if (i < chunks.length - 1) {
        chunks[i] |= 0x80;
      }
      result.push(chunks[i]);
    }
  }
  
  return result;
}

export function decode(bytes: number[]): number[] {
  if (!Array.isArray(bytes)) {
    throw new Error('Input must be array');
  }
  
  const result: number[] = [];
  let current = 0;
  let started = false;
  
  for (let i = 0; i < bytes.length; i++) {
    const byte = bytes[i];
    
    if (byte < 0 || byte > 0xFF) {
      throw new Error('Invalid byte value');
    }
    
    const hasContinuation = (byte & 0x80) !== 0;
    const value = byte & 0x7F;
    
    // Mark that we've started reading a number
    started = true;
    
    // Build the number using multiplication for 32-bit unsigned compatibility
    current = (current * 128) + value;
    
    if (!hasContinuation) {
      // Convert to unsigned 32-bit integer using >>> 0
      result.push(current >>> 0);
      current = 0;
      started = false;
    }
  }
  
  // If we finished loop but still have a number in progress (continuation bit never cleared)
  if (started) {
    throw new Error('Incomplete sequence');
  }
  
  return result;
}

File: solutions/java/variable-length-quantity/1/src/main/java/VariableLengthQuantity.java
import java.util.ArrayList;
import java.util.List;

class VariableLengthQuantity {

    List<String> encode(List<Long> numbers) {
        List<String> result = new ArrayList<>();
        
        for (Long number : numbers) {
            result.addAll(encodeOne(number));
        }
        
        return result;
    }
    
    private List<String> encodeOne(long number) {
        List<String> bytes = new ArrayList<>();
        
        // Handle zero as a special case
        if (number == 0) {
            bytes.add("0x0");
            return bytes;
        }
        
        // Extract 7-bit chunks from LSB to MSB
        List<Integer> chunks = new ArrayList<>();
        long n = number;
        while (n > 0) {
            chunks.add((int)(n & 0x7F));
            n >>>= 7;
        }
        
        // Encode chunks in reverse order (MSB first) with continuation bits
        for (int i = chunks.size() - 1; i >= 0; i--) {
            int chunk = chunks.get(i);
            if (i > 0) {
                // Set continuation bit for all but the last chunk
                chunk |= 0x80;
            }
            bytes.add("0x" + Integer.toHexString(chunk));
        }
        
        return bytes;
    }

    List<String> decode(List<Long> bytes) {
        List<String> result = new ArrayList<>();
        long value = 0;
        boolean inProgress = false;
        
        for (Long byteVal : bytes) {
            int b = byteVal.intValue();
            
            // Extract 7 data bits
            int data = b & 0x7F;
            
            // Shift existing value and add new 7 bits
            value = (value << 7) | data;
            
            // Check if this is the last byte (continuation bit clear)
            if ((b & 0x80) == 0) {
                // Validate that the decoded value fits in 32-bit unsigned
                if (value > 0xFFFFFFFFL) {
                    throw new IllegalArgumentException("Invalid variable-length quantity encoding");
                }
                result.add("0x" + Long.toHexString(value));
                value = 0;
                inProgress = false;
            } else {
                inProgress = true;
            }
        }
        
        // If we ended with an incomplete sequence (last byte had continuation bit set)
        if (inProgress) {
            throw new IllegalArgumentException("Invalid variable-length quantity encoding");
        }
        
        return result;
    }
}